packages/core/src/utils/security/index.ts

```diff
@@ -1,54 +1,86 @@
+  pathToHeader: string[],
+  firstAuthErr: AuthResult,
+) {
+  const flattenedAuthResults = flatten(authResults);
 
-function getAllInvalidSec<R>(invalidSecuritySchemas: Array<Left<AuthErr>>) {
+  return flattenedAuthResults.length === 1
+    ? firstAuthErr
+    : (() => {
+        const wwwAuthenticateHeaders = flattenedAuthResults.map(authResult => {
+          const headers = pipe(
+            authResult,
+            fold((result: AuthResult) => result.headers, (result: AuthResult) => result.headers),
+          );
+
+          return (headers && headers['WWW-Authenticate']) || '';
+        });
+
+        return wwwAuthenticateHeaders.every(identity)
+          ? set(pathToHeader, wwwAuthenticateHeaders.join(', '), firstAuthErr)
+          : firstAuthErr;
+      })();
+}
+
+function getAllInvalidSec<R>(invalidSecuritySchemes: Array<Array<Left<AuthResult>>>) {
   const pathToHeader = ['headers', 'WWW-Authenticate'];
 
-  const firstLeftValue: AuthErr = pipe(
-    invalidSecuritySchemas[0],
-    Either.fold<AuthErr, R, AuthErr>(identity, identity),
+  const firstLeftValue: AuthResult = pipe(
+    invalidSecuritySchemes[0].find(isLeft) || left({}),
+    fold<AuthResult, R, AuthResult>(identity, identity),
   );
 
-  return firstLeftValue.status !== 401 || invalidSecuritySchemas.length === 1
+  return firstLeftValue.status !== 401
     ? firstLeftValue
-    : (() => {
-        const allWWWAuthHeaders = invalidSecuritySchemas.reduce((accumulator: string, currentValue) => {
-          return pipe(
-            currentValue,
-            Either.mapLeft(err => [accumulator, get(err, pathToHeader)].filter(identity).join(', ')),
-            Either.fold(authHeader => authHeader || '', () => ''),
-          );
-        }, '');
-
-        return set(pathToHeader, allWWWAuthHeaders, firstLeftValue);
-      })();
+    : gatherWWWAuthHeader(invalidSecuritySchemes, pathToHeader, firstLeftValue);
 }
 
 export function validateSecurity<R, I>(someInput: I, resource?: R) {
-  const securitySchemas = get(resource, 'security', []);
+  const securitySchemes = get(resource, 'security', []);
 
-  return !securitySchemas.length
+  return !securitySchemes.length
     ? []
     : (() => {
-        const validatedSecuritySchemas = securitySchemas.map((definedSecSchema: SecurityScheme) => {
-          const schemaHandler = securitySchemaHandlers.find(handler => {
-            return handler.test(definedSecSchema);
+        const authResults = securitySchemes.map((securityScheme: SecurityScheme[]) => {
+          const authResult = securityScheme.map((definedSecSchema: SecurityScheme) => {
+            const schemeHandler = securitySchemaHandlers.find(handler => {
+              return handler.test(definedSecSchema);
+            });
+
+            return schemeHandler
+              ? schemeHandler.handle(someInput, definedSecSchema.name, resource)
+              : left({ message: 'We currently do not support this type of security scheme.' });
```

Comment:
@XVincentX hey, do we want to hard error here or let it pass? I think we'd rather keep out the way for things we dont know how to do, and only shout at users when we know they are wrong.